File: src/main/java/org/evertones/persistence/modules/client/ClientDetailsService.java
package org.evertones.persistence.modules.client;

import org.evertones.model.modules.client.ClientDetails;
import org.evertones.model.types.ClientType;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.time.Month;
import java.util.List;

@Service
public class ClientDetailsService {

    private ClientDetailsRepository clientDetailsRepository;

    @Autowired
    public void setClientDetailsRepository(ClientDetailsRepository clientDetailsRepository) {
        this.clientDetailsRepository = clientDetailsRepository;
    }

    public ClientDetails save(ClientDetails clientDetails) {

        if (clientDetails.getCreatedAt() == null) {
            clientDetails.setCreatedAt(LocalDate.now());
        }
        clientDetails.setUpdatedAt(LocalDate.now());

        return clientDetailsRepository.save(clientDetails);
    }

    public List<ClientDetails> findAllByClientType(ClientDetails clientDetails, ClientType clientType) {
        return (List<ClientDetails>) clientDetailsRepository
                .findAll(ClientDetailsSpecification.queryByClient(clientDetails, clientType));
    }

    public List<ClientDetails> findByMonthOfBirth(Month month) {
        return (List<ClientDetails>) clientDetailsRepository.findAll(
                ClientDetailsSpecification.queryClientDetailsBirthday(month)
        );
    }

}

File: src/main/scala/org/evertones/persistence/modules/client/ClientDetailsService.scala
package org.evertones.persistence.modules.client

import java.time.{LocalDate, Month}

import org.evertones.model.modules.client.ClientDetails
import org.evertones.model.types.ClientType
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

import scala.collection.JavaConverters._

@Service
class ClientDetailsService {

    private var clientDetailsRepository: ClientDetailsRepository = _

    @Autowired
    def setClientDetailsRepository(clientDetailsRepository: ClientDetailsRepository) {
        this.clientDetailsRepository = clientDetailsRepository
    }

    def save(clientDetails: ClientDetails): ClientDetails = {

        if (clientDetails.getCreatedAt == null) clientDetails.setCreatedAt(LocalDate.now())
        clientDetails.setUpdatedAt(LocalDate.now())

        clientDetailsRepository.save(clientDetails)
    }

    def findAllByClientType(clientDetails: ClientDetails, clientType: ClientType): List[ClientDetails] = {
        clientDetailsRepository.findAll(
            ClientDetailsSpecification.queryByClient(clientDetails, clientType)).asScala.toList
    }

    def findByMonthOfBirth(month: Month): List[ClientDetails] = {
        clientDetailsRepository.findAll(
            ClientDetailsSpecification.queryClientDetailsBirthday(month)).asScala.toList
    }

}
